src/main/src/main/java/org/geoserver/data/InternationalContentHelper.java

```diff
@@ -0,0 +1,352 @@
+            String[] acceptLanguages,
+            ServiceInfo serviceInfo,
+            List<LayerInfo> layers,
+            List<LayerGroupInfo> groups) {
+        setSupportedLocales(serviceInfo, layers, groups);
+        setRequestedLocales(acceptLanguages);
+        verify();
+    }
+
+    public InternationalContentHelper(
+            String[] acceptLanguages, ServiceInfo serviceInfo, List<ResourceInfo> resources) {
+        setSupportedLocales(serviceInfo, resources);
+        setRequestedLocales(acceptLanguages);
+        verify();
+    }
+
+    public InternationalContentHelper(
+            AcceptLanguagesType acceptLanguagesType,
+            ServiceInfo serviceInfo,
+            List<ResourceInfo> resources) {
+        setSupportedLocales(serviceInfo, resources);
+        setRequestedLocales(acceptLanguagesType);
+        verify();
+    }
+
+    private void setRequestedLocales(AcceptLanguagesType acceptLanguagesType) {
+        if (acceptLanguagesType != null) {
+            EList<String> acceptLanguages = acceptLanguagesType.getLanguage();
+            String[] languagesAr = new String[acceptLanguages.size()];
+            acceptLanguages.toArray(languagesAr);
+            setRequestedLocales(languagesAr);
+        }
+    }
+
+    private void setRequestedLocales(String[] preferredLanguages) {
+        List<String> withoutVariant = new ArrayList<>();
+        if (preferredLanguages != null && preferredLanguages.length > 0) {
+            for (String language : preferredLanguages) {
+                Locale locale = null;
+                if (language.equals("*")) {
+                    this.anyMatch = true;
+                    locale = Locale.getDefault();
+                } else if (language.contains("-")) locale = Locale.forLanguageTag(language);
+                else withoutVariant.add(language);
+                if (locale != null) requestedLocales.add(locale);
+            }
+        }
+        for (String language : withoutVariant) {
+            requestedLocales.addAll(
+                    supportedLocales
+                            .stream()
+                            .filter(l -> l != null && l.getLanguage().equals(language))
+                            .collect(Collectors.toSet()));
+            requestedLocales.add(Locale.forLanguageTag(language));
+        }
+    }
+
+    /**
+     * @param info the ResourceInfo from which retrieve the internationalTitle value
+     * @return the international title found according to the list of requested locales
+     */
+    public String getTitle(ResourceInfo info) {
+        InternationalString internationalString = info.getInternationalTitle();
+        return getFromInternationalString(internationalString);
+    }
+
+    /**
+     * @param info the ResourceInfo from which retrieve the internationalAbstract value
+     * @return the international abstract found according to the list of requested locales
+     */
+    public String getAbstract(ResourceInfo info) {
```

Comment:
The abstract element can be removed from the caps documents if missing (I believe it's not mandatory?) but I don't see code handling that. Also, what happens if the abstract is not specified at all in the config?

Reply:
fixed.